### PR TITLE
IGNITE-12779 : Split implementations of Ignite and IgniteMXBean, make behavior of their active(boolean) different 

### PR DESCRIPTION
**Pull request by Vladsz83**
Behavior changes:

_IgniteMXBean.active(false)_ and _IgniteMXBean.state("inactive")_ now throw an exception if deactivation would clear in-memory data.

Major code changes:

- Implementation of _IgniteMXBean_ extracted from _IgniteKernal_ and put in _IgniteMXBeanImpl_.
- Added IgniteMXBean.state(String, boolean)
- Deprecated _IgniteMXBean.active(boolean)_ and _IgniteMXBean.state(ClusterState)_

Additionally, as discussed, improved comments

/** If {@code true}, cluster deactivation will be forced. */

to more complete descriptions.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 1]
```diff
@@ -3586,6 +3305,26 @@ private void checkNearCacheStarted(IgniteCacheProxy<?, ?> cache) throws IgniteCh
...
     }
 
+    /**
+     * This method allows manually remove the checkpoint with given {@code key}.
+     *
+     * @param key Checkpoint key.
+     * @return {@code true} if specified checkpoint was indeed removed, {@code false}
+     *      otherwise.
+     */
+    boolean removeCheckpoint(String key) {
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: I suggest leaving this method only in the bean.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 2]
```diff
@@ -4618,32 +4357,32 @@ private void registerMetrics() {
...
-
-        reg.register("currentCoordinatorFormatted", this::getCurrentCoordinatorFormatted, String.class,
+        reg.register("uptimeFormatted", this::upTimeFormatted, String.class, UPTIME_FORMATTED_DESC);
+        reg.register("startTimestamp", this::startTimestamp, START_TIMESTAMP_DESC);
+        reg.register("uptime", this::upTime, UPTIME_DESC);
+        reg.register("osInformation", this::osInformation, String.class, OS_INFO_DESC);
+        reg.register("jdkInformation", this::jdkInformation, String.class, JDK_INFO_DESC);
+        reg.register("osUser", this::osUser, String.class, OS_USER_DESC);
+        reg.register("vmName", this::vmName, String.class, VM_NAME_DESC);
+        reg.register("instanceName", this::name, String.class, INSTANCE_NAME_DESC);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: `name` -> `instanceName`

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 3]
```diff
@@ -4652,53 +4391,353 @@ private void registerMetrics() {
...
+        reg.register("failoverSpiFormatted", this::failoverSpiFormatted, String.class,
             FAILOVER_SPI_FORMATTED_DESC);
 
-        reg.register("loadBalancingSpiFormatted", this::getLoadBalancingSpiFormatted, String.class,
+        reg.register("loadBalancingSpiFormatted", this::loadBalancingSpiFormatted, String.class,
             LOAD_BALANCING_SPI_FORMATTED_DESC);
     }
 
+    /**
+     * @return String representation of version of current Ignite instance.
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Can be inlined

File: modules/core/src/main/java/org/apache/ignite/internal/managers/IgniteMBeansManager.java [hunk 4]
```diff
@@ -131,7 +132,7 @@ public void registerAllMBeans(
             return;
 
         // Kernal
-        registerMBean("Kernal", IgniteKernal.class.getSimpleName(), kernal, IgniteMXBean.class);
+        registerMBean("Kernal", IgniteKernal.class.getSimpleName(), new IgniteMXBeanImpl(kernal), IgniteMXBean.class);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/IgniteMBeansManager.java]: Create mbean var, please

File: modules/core/src/main/java/org/apache/ignite/IgniteCluster.java [hunk 5]
```diff
@@ -455,9 +455,6 @@
 
     /**
      * Changes Ignite grid state to active or inactive.
-     * <p>
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/IgniteCluster.java]: Why this changed?

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java (hunk 1 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: > I suggest leaving this method only in the bean.

@NSAmelchev, what do you suggest to do with IgniteKernal.checkClusterState() ?

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override public String getOsUser() {
+        return kernal.osUser();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void printLastErrors() {
+        kernal.context().exceptionRegistry().printErrors(log);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java]: Lets create `printLastErrors` in the kerrnal like `osUser` methods. Log will not be needed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java (hunk 2 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: > `name` -> `instanceName`

We shoulkd not change public api Ignite.name(). And I don't like adding one more function for the same. Makes sense?

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java (hunk 3 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/IgniteCluster.java (hunk 5 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/IgniteCluster.java]: Miss-puss. Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+            throw U.convertException(e);
+        }
+        finally {
+            kernal.context().gateway().readUnlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isRebalanceEnabled() {
+        return kernal.context().cache().context().isRebalanceEnabled();
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java]: Lets use `isRebalanceEnabled` from kernal

Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/managers/IgniteMBeansManager.java (hunk 4 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/IgniteMBeansManager.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+
+        this.log = kernal.context().log(IgniteKernal.class);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean active() {
+        kernal.context().gateway().readLock();
+
+        try {
+            return kernal.context().state().publicApiActiveState(true);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java]: Lets use `kernal.active()`

Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java]: fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java (hunk 2 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Ok, lets use `name()`

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java (hunk 1 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Ok, let's leave it here. And do not move it down in the code.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isRebalanceEnabled() {
+        return kernal.isRebalanceEnabled();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void rebalanceEnabled(boolean rebalanceEnabled) {
+        kernal.context().cache().context().rebalanceEnabled(rebalanceEnabled);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java]: Lets use `kernal.rebalanceEnabled(rebalanceEnabled);`

Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteMXBeanImpl.java]: Fixed.

**Review by NSAmelchev — APPROVED** [review]

**Pull request merged**